### PR TITLE
Change MeshElementGroups from Record to dataclass

### PR DESCRIPTION
**Pull request by alexfikl**
Gave this a try and it seems to not break the world. Yey!

Mostly works towards #283.

### CODE REVIEW

**Comment by alexfikl** [comment]
As a side note, tried converting `Mesh` to be a dataclass too, but the init for that is way too complex to easily translate it.

One big issue was that some arguments like `skip_tests` are theoretically `InitVar` with default values, but dataclasses didn't like that until [recently](https://github.com/python/cpython/pull/20867) :cry:

File: meshmode/mesh/__init__.py [hunk 1]
```diff
@@ -155,98 +156,78 @@ class BTAG_INDUCED_BOUNDARY(BTAG_NO_BOUNDARY):  # noqa: N801
...
-
-        super().__init__(
-            order=order,
-            vertex_indices=vertex_indices,
-            nodes=nodes,
-            unit_nodes=unit_nodes,
-            element_nr_base=element_nr_base, node_nr_base=node_nr_base,
-            **kwargs)
+    order: int
+    vertex_indices: Optional[np.ndarray]
```
Comment (inducer) [inline meshmode/mesh/__init__.py:215]: How come this is marked `Optional`?

File: meshmode/mesh/__init__.py [hunk 2]
```diff
@@ -293,104 +284,162 @@ def __eq__(self, other):
...
+            of an element in this group.
+        """
+        raise NotImplementedError
+
+    def vertex_unit_coordinates(self) -> np.ndarray:
+        """
+        :returns: an array of shape ``(nfaces, dim)`` with the unit
+            coordinates of each vertex.
+        """
+        raise NotImplementedError
```
Comment (inducer) [inline meshmode/mesh/__init__.py:304]: Make this an ABC?

File: meshmode/mesh/__init__.py [hunk 3]
```diff
@@ -293,104 +284,162 @@ def __eq__(self, other):
...
-            :func:`modepy.edge_clustered_nodes_for_space` are assumed.
-            These must be in unit coordinates as defined in :mod:`modepy`.
-        :arg dim: only used if *unit_nodes* is *None*, to get
-            the default unit nodes.
-
-        Do not supply *element_nr_base* and *node_nr_base*, they will be
-        automatically assigned.
-        """
+    shape_cls: ClassVar[Type[mp.Shape]] = mp.Shape
+    shape_dim: Optional[int] = None
```
Comment (inducer) [inline meshmode/mesh/__init__.py:314]: Why optional?

File: meshmode/mesh/__init__.py [hunk 4]
```diff
@@ -293,104 +284,162 @@ def __eq__(self, other):
...
+
+    @property
+    def _modepy_space(self):
+        return self.space
 
     @property
-    @memoize_method
     def _modepy_faces(self):
-        return mp.faces_for_shape(self._modepy_shape)
+        return self.faces
```
Comment (inducer) [inline meshmode/mesh/__init__.py:372]: I guess I'm on board with un-underscoring these, but then these should be deprecated.

File: meshmode/mesh/__init__.py [hunk 5]
```diff
@@ -412,6 +461,18 @@ class NodalAdjacency(Record):
     .. automethod:: __ne__
     """
 
+    neighbors_starts: np.ndarray
+    neighbors: np.ndarray
+
+    def copy(self, **kwargs: Any) -> "NodalAdjacency":
+        from warnings import warn
+        warn(f"{type(self).__name__}.copy is deprecated and will be removed in 2022."
```
Comment (inducer) [inline meshmode/mesh/__init__.py:469]:
```suggestion
        warn(f"{type(self).__name__}.copy is deprecated and will be removed in July 2022."
```

**Review by inducer — COMMENTED** [review]
Thanks for working on this! Looks good except for these (hopefully) minor wrinkles.

**Comment by inducer** [comment]
Unsubscribing... @-mention or request review once it's ready for a look or needs attention.

File: meshmode/mesh/__init__.py (hunk 1 above)
Comment (alexfikl) [reply meshmode/mesh/__init__.py:215]: We made vertices optional a while back in here
https://gitlab.tiker.net/inducer/meshmode/-/merge_requests/66

File: meshmode/mesh/__init__.py (hunk 2 above)
Comment (alexfikl) [reply meshmode/mesh/__init__.py]: Done in 4b41215.

File: meshmode/mesh/__init__.py (hunk 4 above)
Comment (alexfikl) [reply meshmode/mesh/__init__.py]: Un-un-underscored them in 0807521. There was no reason for the name change and un-underscoring in the first place.

File: meshmode/mesh/__init__.py (hunk 3 above)
Comment (alexfikl) [reply meshmode/mesh/__init__.py]: Reworked this a bit in 0807521. It seems to work and hopefully make this clearer.

**Comment by alexfikl** [comment]
This should be ready for another look. I still don't like the big `__post_init__` in the modepy element groups, but it's the same as before.

File: meshmode/mesh/__init__.py [hunk 6]
```diff
@@ -293,104 +297,133 @@ def __eq__(self, other):
...
+    dim: Optional[int] = None
+    unit_nodes: Optional[np.ndarray] = None
+
+    _modepy_shape_cls: ClassVar[Type[mp.Shape]] = mp.Shape
+    _modepy_shape: mp.Shape = field(default=None, init=False, repr=False)
+    _modepy_space: mp.FunctionSpace = field(default=None, init=False, repr=False)
 
-        if unit_nodes is None:
-            unit_nodes = mp.edge_clustered_nodes_for_space(space, shape)
+    def __post_init__(self):
```
[the remark this answers is not in the document]
Comment (alexfikl) [reply meshmode/mesh/__init__.py:343]: @inducer Hm, not quite sure how to implement that in a vaguely backwards compatible way though. Any suggestions / preferences?

[the remark this answers is not in the document]
Comment (inducer) [reply meshmode/mesh/__init__.py:343]: Maybe make a no-op factory function for now and raise a `DeprecationWarning` for construction that doesn't go through the function?

[the remark this answers is not in the document]
Comment (alexfikl) [reply meshmode/mesh/__init__.py:356]: Tried it in 2ecce05. Let me know if that's more or less what you had in mind.

**Review by inducer — APPROVED** [review]
Thanks! LGTM, queued to go in after #299.

**Pull request merged**